### PR TITLE
desktop/windowRule: add active_border and inactive_border rules

### PR DESCRIPTION
**Pull request by RommelRivera**
#### Describe your PR, what does it fix/add?

This PR aims to resolve issues with configuring inactive border colors alongside active border colors for window rules, as pointed out in discussions #14030 and #10067.

When attempting to add a solid color for active and inactive borders, using the string representation specifically, because with lua tables it also appears to not parse correctly, both border colors are set correctly. e.g.:

> border_color = "rgba(ff0000ee) rgba(00ff00ee)"

However, if a gradient is defined for both, or just one of the border colors, unusual behavior is seen. e.g.:

> border_color =  "rgba(ff0000ee) rgba(00ff00ee) 45deg rgba(ffff00ee) rgba(00ffffee) 45deg"

Some of the behavior observed includes border colors taking a little while to update, a solid gray color instead of the defined color/gradient, flickering colors, and app slowdowns (though this may just be due to my PC specs, take this effect with a grain of salt), I've attached a video showing some of this happening, discussion #14030 also includes a video.

https://github.com/user-attachments/assets/e87ce134-8085-4673-b59c-f2388e8d4ea2

The way this issue is resolved is by simply adding a line to run the `updateColorsOk()` function for the inactive color inside of the function that parses the border_color window rule.

This also allows to set only an inactive border color by writing a string like so:

> border_color = "0deg rgba(ffffffee) rgba(000000ee) 45deg"

Although admittedly this is quite hacky, so I also added the following two new window rules: `active_border` and `inactive_border`, which work the same as the core configs, essentially, what I've done is copy the function that parses the `border_color` rule, once for each window state, and just set the active color in one function and the inactive in another, which lets us write for these rules in the same way one would for the global active and inactive border colors. These options only allow lua tables and not a string representation as the original `border_color` option allows.

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)

In order to allow for only specifying the inactive border color, I would have had to check if the list of elements inside of the border color object was greater than 0, so I decided to also change the definition of the active color to an `std::optional<Config::CGradientValueData>`, as far as I was able to tell, this should not break anything, but if this is a weird fix, I can go back to checking the size of the object's elements array.

The reason for this is that if we assume that the active border color is set when it isn't, because the active color was not configured in the window rule, it would show the active border as a dark gray color instead of the globally defined active border color.

#### Is it ready for merging, or does it need work?

As far as I can tell, it is ready for merging, only 4 tests didn't pass on my machine, but I had to run them outside of hyprland bc running the tests crashed hyprland for me, so that may be why the tests didn't pass.

### CODE REVIEW

**Comment by vaxerski** [comment]
I think a better solution is to scrap border_color taking 2 colors and simply rely on 2 window rules.

**Comment by RommelRivera** [comment]
Done, `border_color` is no more @vaxerski

**Comment by vaxerski** [comment]
that is the opposite of what I said. Have border_color override both active and inactive with one color, and rely on two window rules (one for active, one for inactive) to apply.

**Comment by RommelRivera** [comment]
I'm kinda confused, what do you mean by using `active_border` and `inactive_border` to apply the colors? Could you give me an example?

Do you mean like, `border_color` sets both colors at once and `active_border` then overrides the active color? Or that `active_border` and `inactive_border` depend on `border_color` in some way?

**Comment by vaxerski** [comment]
Window rule 1:
 - match: something
 - match: active yes
 - set border_color "my_active_color"
 
Window rule 2:
 - match: something
 - match: active no
 - set border_color "my_inactive_color"

**Comment by RommelRivera** [comment]
In that case, how would I set both active and inactive colors? It seems that with your example, only one can be set at once.

**Comment by vaxerski** [comment]
you just make two window rules?

**Comment by RommelRivera** [comment]
Yeah, though two window rules for that seems kinda weird to me, but I'll go ahead and implement it

**Comment by vaxerski** [comment]
I think it makes more sense because it's much more explicit. I don't like how we have currently like 3 ways a border color can change, and it never satisfies users. I was thinking of removing border color settings for inactive altogether.

**Comment by RommelRivera** [comment]
Currently, the following will set Dolphin's active and inactive borders completely separate from the main active_border and inactive_border global configs:

```
hl.window_rule({
	name = "Dolphin",
	match = { class = 'org.kde.dolphin', focus = true },
	border_color = { colors = { "rgba(ff0000ee)", "rgba(00ff00ee)", "rgba(0000ffee)" }, angle = 45 }
})

hl.window_rule({
	name = "Dolphin_2",
	match = { class = 'org.kde.dolphin', focus = false },
	border_color = { colors = { "rgba(00ffffee)", "rgba(ff00ffee)", "rgba(ffff00ee)" }, angle = 45 }
})
```

**Comment by vaxerski** [comment]
this is good, but we should preserve compatibility with legacy configs.

**Comment by RommelRivera** [comment]
Legacy configs _shouldn't_ break right now, assuming I haven't missed some edge case in the tests.

**Comment by vaxerski** [comment]
conflict needs to be resolved then we can run tests and see lol

**Comment by RommelRivera** [comment]
Uh, don't know what I did to get it closed automatically, all I did was sync the fork to main for rebasing my changes, anything I can do at the moment?

**Comment by vaxerski** [comment]
I vouched you and reopened :)

**Comment by RommelRivera** [comment]
Fuck me somehow I did it again but at least i can reopen lol, I have already rebased the latest commits from the main branch with my changes, and double checked to see that I didn't overwrite anything made by someone else.

**Comment by vaxerski** [comment]
@codex review

File: hyprtester/src/tests/main/window.cpp [hunk 1]
```diff
@@ -2037,3 +2021,161 @@ TEST_CASE(specialWorkspacePinnedWindowUnmapFallback) {
...
+
+}
+
+TEST_CASE(windowRuleBorderColorsFocus) {
+    // Setting global border colors to check against
+    OK(getFromSocket("/eval hl.config({ general = { col = { active_border = { colors = { 'rgba(00ff00ff)', 'rgba(ffffffff)' }, angle = 45 } } } })"));
+    OK(getFromSocket("/eval hl.config({ general = { col = { inactive_border = { colors = { 'rgba(ff0000ff)', 'rgba(ffffffff)' }, angle = 45 } } } })"));
+
+    // Testing setting only the 'active_border' option
+    OK(getFromSocket("/eval hl.window_rule({ name = 'active-border-kitty', match = { class = 'active_border_kitty', focus = true }, border_color = { colors = {'rgba(ff00ffff)', 'rgba(000000ff)' }, angle = 90 } })"));
```
[the remark this answers is not in the document]
Comment (RommelRivera) [reply hyprtester/src/tests/main/window.cpp]: I've just updated the comments to remove references to the non-existent active_border and inactive_border rules

[the remark this answers is not in the document]
Comment (RommelRivera) [reply hyprtester/src/tests/main/window.cpp]: @codex address the feedback

[the remark this answers is not in the document]
Comment (erstarr) [reply hyprtester/src/tests/main/window.cpp]: > @codex address the feedback

That is against the AI policy. AI cannot interact with developer facing things including reviews. If you want your AI to do the work for you; that's allowed only if it's declared clearly.

Also if you are doing this i'm guessing you already used AI in this MR. Disclose it.

[the remark this answers is not in the document]
Comment (RommelRivera) [reply hyprtester/src/tests/main/window.cpp]: I haven't used AI at all for this, I was just trying to see if it would respond to the latest commit since vaxry asked it to review the PR in the first place

**Review by vaxerski — DISMISSED** [review]
wiki mr please

**Comment by RommelRivera** [comment]
:saluting_face: on it

**Comment by vaxerski** [comment]
clang-format pls

**Review by vaxerski — APPROVED** [review]
